pkg/gce-pd-csi-driver/node_test.go

```diff
@@ -328,6 +341,87 @@ func TestNodeStageVolume(t *testing.T) {
 	}
 }
 
+func TestNodeExpandVolume(t *testing.T) {
+	// TODO: Add tests/functionality for non-existant volume
+	var resizedBytes int64 = 2000000000
+	volumeID := "project/test001/zones/c1/disks/testDisk"
+	testCases := []struct {
+		name         string
+		req          *csi.NodeExpandVolumeRequest
+		blockDevice  bool
+		expRespBytes int64
+		expErrCode   codes.Code
+	}{
+		{
+			name: "ext4 fs expand",
+			req: &csi.NodeExpandVolumeRequest{
+				VolumeId:   volumeID,
+				VolumePath: "some-path",
+				CapacityRange: &csi.CapacityRange{
+					RequiredBytes: resizedBytes,
+				},
+			},
+			blockDevice:  false,
+			expRespBytes: resizedBytes,
+		},
+		{
+			name: "block device expand",
+			req: &csi.NodeExpandVolumeRequest{
+				VolumeId:   volumeID,
+				VolumePath: "some-path",
+				CapacityRange: &csi.CapacityRange{
+					RequiredBytes: resizedBytes,
+				},
+			},
+			blockDevice: true,
```

Comment:
does expRespBytes need to be set?

If for some reason, nodeexpand got called before controllerexpand succeeded, do we want to blindly return success?

Reply:
I was attempting to interpret the spec (specifically the `OPTIONAL` bit):
> // The capacity of the volume in bytes. This field is OPTIONAL.
  int64 capacity_bytes = 1;

In that I think since after the `NodeExpandVolume` the `volume` which for `NodeExpandVolume` I feel refers to the fs, the block volume has no `capacity_bytes`. I could be easily convinced otherwise though, this is just my best effort guess

Reply:
cc @gnufied for spec interpretation and why is capacity in the response optional?  Does the resizer use the returned capacity to set the PVC capacity? I don't think it specifically refers to fs/block anywhere in NodeExpandVolume. fs was used as an example.

Reply:
`capacity_bytes` in response and `capacity_range` in request is optional for `NodeExpandVolume` and the reason this is optional is because a CO may choose to expand the volume to full available size for fs expansion (if applicable). Obviously knowing whether to set `capacity_range` or not depends on ability of caller to determine what kind of volume it is dealing with when making `NodeExpandVolume` call. For a plugin author though - it always a good idea to set `capacity_bytes` in response whenever possible.

Currently we aren't using the return size from this call for setting `pvc.Status.Capacity` but another CO or a different version of k8s may use the returned size.

Reply:
One potential problem with setting return value of this call to `PVC.Status.Capacity` is - after formatting with a FS, the result disk size might be smaller than what user requested but we do not want to trigger another resize of underlying PV for handling such a case.

Reply:
got it. Setting `response` size to equal `request` if successful resize finished. We also ignore the capacity range in the request for now but I've created a tracking issue for that